File: scripts/telegram-chat-check.test.py
"""Проверка разбора чатов — без обращения к Telegram.

Разбор решает, что Дарья прочитает в журнале, поэтому ветки проверяются
на подставных ответах: ошибка в тексте стоит потерянного дня.
"""

import importlib.util
import io
import os
import sys
import unittest.mock
from pathlib import Path

spec = importlib.util.spec_from_file_location(
    'chat_check', Path(__file__).with_name('telegram-chat-check.py')
)
chat_check = importlib.util.module_from_spec(spec)
spec.loader.exec_module(chat_check)

BOT = {'ok': True, 'result': {'username': 'testbot', 'first_name': 'Проверка'}}
DENIED = {'ok': False, 'description': 'Bad Request: chat not found'}


def run(env, answers):
    """Прогнать разбор, подставив ответы Telegram по (метод, chat_id)."""
    def fake_call(_token, method, **params):
        if method == 'getMe':
            return BOT
        return answers.get(params.get('chat_id'), DENIED)

    out = io.StringIO()
    with unittest.mock.patch.object(chat_check, 'call', fake_call), \
            unittest.mock.patch.dict(os.environ, env, clear=True), \
            unittest.mock.patch.object(sys, 'stdout', out):
        code = chat_check.main()
    return code, out.getvalue()


def check(name, condition):
    if not condition:
        print(f'ПРОВАЛ: {name}')
        sys.exit(1)
    print(f'  ✓ {name}')


base = {'TELEGRAM_BOT_TOKEN': 'x'}

# 1. Рабочий чат — проверка зелёная.
code, log = run(
    {**base, 'TELEGRAM_MANAGER_CHAT_IDS': '-1001111'},
    {'-1001111': {'ok': True, 'result': {'title': 'Менеджеры', 'type': 'supergroup'}}},
)
check('рабочий чат: код 0', code == 0)
check('рабочий чат: название видно', '«Менеджеры»' in log)
check('рабочий чат: id в журнал не попал', '-1001111' not in log)

# 2. Группа выросла в супергруппу — причина названа прямо.
code, log = run(
    {**base, 'TELEGRAM_MANAGER_CHAT_IDS': '-1111'},
    {'-1001111': {'ok': True, 'result': {'title': 'Менеджеры', 'type': 'supergroup'}}},
)
check('повышение до супергруппы: код 1', code == 1)
check('повышение до супергруппы: причина названа', 'повысили до супергруппы' in log)
check('повышение до супергруппы: новый id не напечатан',
      log.count('-1001111') == 1 and '::add-mask::-1001111' in log)

# 3. Мусор вместо id.
code, log = run({**base, 'TELEGRAM_MANAGER_CHAT_IDS': 'https://t.me/chat'}, {})
check('мусор в секрете: код 1', code == 1)
check('мусор в секрете: сказано, что это не id', 'НЕ ПОХОЖЕ НА ID' in log)

# 4. Верный по виду id, но бота в чате нет.
code, log = run({**base, 'TELEGRAM_MANAGER_CHAT_IDS': '-1002222'}, {})
check('бот не в чате: код 1', code == 1)
check('бот не в чате: назван логин бота', '@testbot' in log)

# 5. Личный чат без «Start».
code, log = run({**base, 'TELEGRAM_MANAGER_CHAT_IDS': '555'}, {})
check('личный чат: подсказка про Start', 'Start' in log)

# 6. Часть чатов жива — проверка зелёная, но предупреждает.
code, log = run(
    {**base, 'TELEGRAM_MANAGER_CHAT_IDS': '-1001111,-1002222'},
    {'-1001111': {'ok': True, 'result': {'title': 'Менеджеры', 'type': 'supergroup'}}},
)
check('часть чатов жива: код 0', code == 0)
check('часть чатов жива: есть предупреждение', 'дойдут, но не всем' in log)

print('Разбор чатов: все проверки пройдены')

File: scripts/telegram-chat-check.test.mjs
import assert from 'node:assert/strict';
import { describe, it } from 'node:test';

import { runTelegramChatCheck } from './telegram-chat-check.mjs';

const BOT = { ok: true, result: { username: 'testbot', first_name: 'Проверка' } };
const DENIED = { ok: false, description: 'Bad Request: chat not found' };

const run = async (managerChatIds, answers = {}) => {
    const lines = [];
    const call = async (_token, method, params = {}) => {
        if (method === 'getMe') return BOT;
        return answers[params.chat_id] ?? DENIED;
    };
    const code = await runTelegramChatCheck({
        env: {
            TELEGRAM_BOT_TOKEN: 'test-token',
            TELEGRAM_MANAGER_CHAT_IDS: managerChatIds,
        },
        call,
        write: (line) => lines.push(line),
    });
    return { code, output: lines.join('\n') };
};

describe('проверка доступности чатов Telegram', () => {
    it('понятно сообщает об отсутствующем токене', async () => {
        const lines = [];
        const code = await runTelegramChatCheck({
            env: { TELEGRAM_MANAGER_CHAT_IDS: '-1001111' },
            write: (line) => lines.push(line),
        });
        assert.equal(code, 1);
        assert.match(lines.join('\n'), /Не задан секрет TELEGRAM_BOT_TOKEN/);
    });

    it('понятно сообщает о пустом списке чатов', async () => {
        const lines = [];
        const code = await runTelegramChatCheck({
            env: { TELEGRAM_BOT_TOKEN: 'test-token', TELEGRAM_MANAGER_CHAT_IDS: '   ' },
            write: (line) => lines.push(line),
        });
        assert.equal(code, 1);
        assert.match(lines.join('\n'), /Не задан секрет TELEGRAM_MANAGER_CHAT_IDS/);
    });

    it('завершается успешно для рабочего чата и не печатает его id', async () => {
        const id = '-1001111';
        const result = await run(id, {
            [id]: { ok: true, result: { title: 'Менеджеры', type: 'supergroup' } },
        });
        assert.equal(result.code, 0);
        assert.match(result.output, /«Менеджеры»/);
        assert.doesNotMatch(result.output, new RegExp(id));
    });

    it('распознаёт группу, повышенную до супергруппы, без раскрытия обоих id', async () => {
        const oldId = '-1111';
        const newId = '-1001111';
        const result = await run(oldId, {
            [newId]: { ok: true, result: { title: 'Менеджеры', type: 'supergroup' } },
        });
        assert.equal(result.code, 1);
        assert.match(result.output, /повысили до супергруппы/);
        assert.doesNotMatch(result.output, new RegExp(oldId));
        assert.doesNotMatch(result.output, new RegExp(newId));
    });

    it('останавливается до обращения к чату, если в секрете ссылка', async () => {
        const result = await run('https://t.me/chat');
        assert.equal(result.code, 1);
        assert.match(result.output, /в секрете не номер чата/);
        assert.doesNotMatch(result.output, /https|t\.me/);
    });

    it('называет бота, которого надо добавить в недоступный чат', async () => {
        const result = await run('-1002222');
        assert.equal(result.code, 1);
        assert.match(result.output, /@testbot/);
        assert.doesNotMatch(result.output, /-1002222/);
    });

    it('объясняет, что для личного чата надо нажать Start', async () => {
        const result = await run('555');
        assert.match(result.output, /Start/);
        assert.doesNotMatch(result.output, /555/);
    });

    it('сохраняет доставку в рабочий чат, если другая запись испорчена', async () => {
        const id = '-1001111';
        const result = await run(`${id},https://t.me/chat`, {
            [id]: { ok: true, result: { title: 'Менеджеры', type: 'supergroup' } },
        });
        assert.equal(result.code, 0);
        assert.match(result.output, /дойдут только в доступные чаты/);
        assert.doesNotMatch(result.output, new RegExp(id));
        assert.doesNotMatch(result.output, /https|t\.me/);
    });
});
